### PR TITLE
fix(animations): `animateChild` shouldn't trigger animation on disabled element

### LINKED ISSUE #35558: `animateChild` invoke disabled child animation in group
# 🐞 bug report

### Affected Package
The issue is caused by package `@angular/animations`

### Is this a regression?
I don't know

### Description
I have group animations. Child animation is disabled on small screen, but child animation is called  by `animateChild` despite being disabled.

```
 trigger('parentAnimation', [
      state('in', style({opacity: 1})),
      transition(':enter',  group([
        query('.main', [style({opacity: 0}), animate('2s ease')]),
        query('@*', animateChild())
      ])),
      ...
```

```
<div @parentAnimation *ngIf="display">
  <p class="main">Should animate</p>
  <p @childAnimation [@.disabled]="true">
    Shouldn't animates
  </p>
</div>
```

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/animate-child

In Chrome and FF two animations are showing. 
In Safari no animations.

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 8.3.20
Node: 12.10.0
OS: darwin x64
Angular: 8.2.14
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router, upgrade

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.803.20
@angular-devkit/build-angular     0.803.20
@angular-devkit/build-optimizer   0.803.20
@angular-devkit/build-webpack     0.803.20
@angular-devkit/core              8.3.8
@angular-devkit/schematics        8.3.20
@angular/cdk                      8.2.3
@angular/cli                      8.3.20
@ngtools/webpack                  8.3.8
@schematics/angular               8.3.20
@schematics/update                0.803.20
rxjs                              6.5.3
typescript                        3.5.3
webpack                           4.41.2
</code></pre>

### PR DESCRIPTION
**Pull request by Zuckjet**
…ations

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
Fixes [35558](https://github.com/angular/angular/issues/35558)

### CODE REVIEW

**Comment by Zuckjet** [comment]
@googlebot I fixed it.

**Comment by Zuckjet** [comment]
@googlebot I fixed it.

**Comment by Zuckjet** [comment]
Before continue this PR, I have some problems.

- _ci/circleci: test_aio_local_ failed with message:

> FAIL: Commit dfd3af522ce0c85d39f030036696d3b825c8b904 uncompressed main-es2015 exceeded expected size by 500 bytes or >1% 
this error is strange to me, and seems is not my code error

- I think this PR should add some tests, but when I read the test file ,it seems has some typo error:
https://github.com/angular/angular/blob/4481bebb38bdc93f2214f354deee7c5a174e05cf/packages/core/test/animation/animation_query_integration_spec.ts#L3207-L3229

**Comment by Zuckjet** [comment]
just friendly ping @mhevery @matsko

**Review by mhevery — APPROVED** [review]

**Comment by mhevery** [comment]
[presubmit](http://test/OCL:342720142:BASE:342720196:1605565360009:2ceace9e)

**Review by mhevery — APPROVED** [review]
reviewed-for: global-approvers

**Comment by AndrewKushnir** [comment]
Restarting a [presubmit](http://test/OCL:343170693:BASE:343170702:1605742692323:92cce7d) since the previous one contained a failing target.

**Comment by mhevery** [comment]
The failures in `google3` seem to be real (both VE and Ivy fail in the same way). It seems that some styling is not being applied resulting in text being shifted few pixels. Needs investigation

**Comment by Zuckjet** [comment]
Thanks for review, it seems I cannot find out what failure happened in `google3`, so cannot do anything help.
If this PR(it self) need changes, I am glad to help.

**Comment by mhevery** [comment]
> so cannot do anything help

Right so this is blocked on investigating what is going on in `google3`. This will likely be a while.

File: packages/core/test/animation/animation_query_integration_spec.ts [hunk 1]
```diff
@@ -3302,6 +3302,63 @@ describe('animation query tests', function() {
            expect(p2.duration).toEqual(1000);
            expect(p2.element.classList.contains('parent')).toBeTrue();
          });
+
+      it('should disable the animation for the element that they are disabled on even if a parent animation query and animate sub animations',
```
Comment (jessicajaniuk) [inline packages/core/test/animation/animation_query_integration_spec.ts:3306]: This is a confusing sentence that makes it hard to follow the test's intent. I think this option is clearer.
```suggestion
      it('should disable the animation for the given element regardless of existing parent element animation queries or child element animation queries',
```

**Review by jessicajaniuk — APPROVED** [review]
LGTM with one suggestion for clarification on the unit test name

File: packages/core/test/animation/animation_query_integration_spec.ts (hunk 1 above)
Comment (Zuckjet) [reply packages/core/test/animation/animation_query_integration_spec.ts]: Thanks, have fixed.

**Comment by Zuckjet** [comment]
The cr/circleci error seems is not relevant with this PR.

**Comment by jessicajaniuk** [comment]
I've restarted the CI target and am running a google3 presubmit.

File: packages/animations/browser/src/render/transition_animation_engine.ts [hunk 2]
```diff
@@ -1033,7 +1033,10 @@ export class TransitionAnimationEngine {
         // instead stretch the first keyframe gap up until the animation starts. The
         // reason this is important is to prevent extra initialization styles from being
         // required by the user in the animation.
-        instruction.timelines.forEach(tl => tl.stretchStartingKeyframe = true);
+        instruction.timelines = instruction.timelines.filter(tl => {
```
Comment (crisbeto) [inline packages/animations/browser/src/render/transition_animation_engine.ts:1036]: Overall the changes look good, but this `filter` with a side effect looks weird since filter predicate functions are supposed to be pure. I'd recommend doing this instead:

```
const timelines = [];

instruction.timelines.forEach(tl => {
  tl.stretchStartingKeyframe = true;

  if (!this.disabledNodes.has(tl.element)) {
    timelines.push(tl);
  }
});

instruction.timelines = timelines;
```

Comment (Zuckjet) [reply packages/animations/browser/src/render/transition_animation_engine.ts]: Thanks, have updated code

**Review by crisbeto — APPROVED** [review]
LGTM

**Comment by Zuckjet** [comment]
friendly ping @IgorMinar

**Comment by jessicajaniuk** [comment]
@Zuckjet Looks like there's a code conflict here. Can you rebase this? I think we should be able to land it. Once you rebase, I'll re-run a presubmit and we should be green.

**Comment by Zuckjet** [comment]
> @Zuckjet Looks like there's a code conflict here. Can you rebase this? I think we should be able to land it. Once you rebase, I'll re-run a presubmit and we should be green.

have fixed the conflict.

**Comment by jessicajaniuk** [comment]
Everything looks good. Thanks @Zuckjet!

**Comment by alxhub** [comment]
This PR was merged into the repository by commit bab7ed35418173714509e06ba8b092aff28fff89.
